running "exit" fails with FileNotFoundError

The first two examples (testing exit) fail on my machine. The others succeed as expected.

```
❱ cat exit-test.py
def test_assert_good_exitcode(shell):

    ret = shell.run("exit", "0")
    assert ret.returncode == 0
```

Executing:

```
❱ pytest exit-test.py
===================================================================== test session starts =====================================================================
platform linux -- Python 3.10.6, pytest-7.1.2, pluggy-1.0.0
rootdir: /tmp/pytest-shell-utilities
plugins: skip-markers-1.3.0, shell-utilities-1.6.0, helpers-namespace-2021.12.29
collected 1 item                                                                                                                                              

exit-test.py F                                                                                                                                          [100%]

========================================================================== FAILURES ===========================================================================
__________________________________________________________________ test_assert_good_exitcode __________________________________________________________________

shell = Subprocess(cwd=PosixPath('/tmp/pytest-shell-utilities'), slow_stop=True, timeout=None)

    def test_assert_good_exitcode(shell):
    
>       ret = shell.run("exit", "0")

exit-test.py:3: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/home/aqw/.venvs/onyo/lib/python3.10/site-packages/pytestshellutils/shell.py:490: in run
    self.impl.run(*args, env=env, **kwargs)
/home/aqw/.venvs/onyo/lib/python3.10/site-packages/pytestshellutils/shell.py:323: in run
    return self.init_terminal(cmdline, env=env)
/home/aqw/.venvs/onyo/lib/python3.10/site-packages/pytestshellutils/shell.py:158: in init_terminal
    self._terminal = subprocess.Popen(
/usr/lib/python3.10/subprocess.py:969: in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <Popen: returncode: 255 args: ['exit', '0']>, args = ['exit', '0'], executable = b'exit', preexec_fn = None, close_fds = True, pass_fds = ()
cwd = '/tmp/pytest-shell-utilities'
env = {'COLORTERM': 'truecolor', 'DBUS_SESSION_BUS_ADDRESS': 'unix:path=/run/user/1000/bus', 'DISPLAY': ':0', 'DOTFILES': '/home/aqw/.dotfiles', ...}
startupinfo = None, creationflags = 0, shell = False, p2cread = -1, p2cwrite = -1, c2pread = -1, c2pwrite = 11, errread = -1, errwrite = 12
restore_signals = True, gid = None, gids = None, uid = None, umask = -1, start_new_session = False

    def _execute_child(self, args, executable, preexec_fn, close_fds,
                       pass_fds, cwd, env,
                       startupinfo, creationflags, shell,
                       p2cread, p2cwrite,
                       c2pread, c2pwrite,
                       errread, errwrite,
                       restore_signals,
                       gid, gids, uid, umask,
                       start_new_session):
        """Execute program (POSIX version)"""
    
        if isinstance(args, (str, bytes)):
            args = [args]
        elif isinstance(args, os.PathLike):
            if shell:
                raise TypeError('path-like args is not allowed when '
                                'shell is true')
            args = [args]
        else:
            args = list(args)
    
        if shell:
            # On Android the default shell is at '/system/bin/sh'.
            unix_shell = ('/system/bin/sh' if
                      hasattr(sys, 'getandroidapilevel') else '/bin/sh')
            args = [unix_shell, "-c"] + args
            if executable:
                args[0] = executable
    
        if executable is None:
            executable = args[0]
    
        sys.audit("subprocess.Popen", executable, args, cwd, env)
    
        if (_USE_POSIX_SPAWN
                and os.path.dirname(executable)
                and preexec_fn is None
                and not close_fds
                and not pass_fds
                and cwd is None
                and (p2cread == -1 or p2cread > 2)
                and (c2pwrite == -1 or c2pwrite > 2)
                and (errwrite == -1 or errwrite > 2)
                and not start_new_session
                and gid is None
                and gids is None
                and uid is None
                and umask < 0):
            self._posix_spawn(args, executable, env, restore_signals,
                              p2cread, p2cwrite,
                              c2pread, c2pwrite,
                              errread, errwrite)
            return
    
        orig_executable = executable
    
        # For transferring possible exec failure from child to parent.
        # Data format: "exception name:hex errno:description"
        # Pickle is not used; it is complex and involves memory allocation.
        errpipe_read, errpipe_write = os.pipe()
        # errpipe_write must not be in the standard io 0, 1, or 2 fd range.
        low_fds_to_close = []
        while errpipe_write < 3:
            low_fds_to_close.append(errpipe_write)
            errpipe_write = os.dup(errpipe_write)
        for low_fd in low_fds_to_close:
            os.close(low_fd)
        try:
            try:
                # We must avoid complex work that could involve
                # malloc or free in the child process to avoid
                # potential deadlocks, thus we do all this here.
                # and pass it to fork_exec()
    
                if env is not None:
                    env_list = []
                    for k, v in env.items():
                        k = os.fsencode(k)
                        if b'=' in k:
                            raise ValueError("illegal environment variable name")
                        env_list.append(k + b'=' + os.fsencode(v))
                else:
                    env_list = None  # Use execv instead of execve.
                executable = os.fsencode(executable)
                if os.path.dirname(executable):
                    executable_list = (executable,)
                else:
                    # This matches the behavior of os._execvpe().
                    executable_list = tuple(
                        os.path.join(os.fsencode(dir), executable)
                        for dir in os.get_exec_path(env))
                fds_to_keep = set(pass_fds)
                fds_to_keep.add(errpipe_write)
                self.pid = _posixsubprocess.fork_exec(
                        args, executable_list,
                        close_fds, tuple(sorted(map(int, fds_to_keep))),
                        cwd, env_list,
                        p2cread, p2cwrite, c2pread, c2pwrite,
                        errread, errwrite,
                        errpipe_read, errpipe_write,
                        restore_signals, start_new_session,
                        gid, gids, uid, umask,
                        preexec_fn)
                self._child_created = True
            finally:
                # be sure the FD is closed no matter what
                os.close(errpipe_write)
    
            self._close_pipe_fds(p2cread, p2cwrite,
                                 c2pread, c2pwrite,
                                 errread, errwrite)
    
            # Wait for exec to fail or succeed; possibly raising an
            # exception (limited in size)
            errpipe_data = bytearray()
            while True:
                part = os.read(errpipe_read, 50000)
                errpipe_data += part
                if not part or len(errpipe_data) > 50000:
                    break
        finally:
            # be sure the FD is closed no matter what
            os.close(errpipe_read)
    
        if errpipe_data:
            try:
                pid, sts = os.waitpid(self.pid, 0)
                if pid == self.pid:
                    self._handle_exitstatus(sts)
                else:
                    self.returncode = sys.maxsize
            except ChildProcessError:
                pass
    
            try:
                exception_name, hex_errno, err_msg = (
                        errpipe_data.split(b':', 2))
                # The encoding here should match the encoding
                # written in by the subprocess implementations
                # like _posixsubprocess
                err_msg = err_msg.decode()
            except ValueError:
                exception_name = b'SubprocessError'
                hex_errno = b'0'
                err_msg = 'Bad exception data from child: {!r}'.format(
                              bytes(errpipe_data))
            child_exception_type = getattr(
                    builtins, exception_name.decode('ascii'),
                    SubprocessError)
            if issubclass(child_exception_type, OSError) and hex_errno:
                errno_num = int(hex_errno, 16)
                child_exec_never_called = (err_msg == "noexec")
                if child_exec_never_called:
                    err_msg = ""
                    # The error must be from chdir(cwd).
                    err_filename = cwd
                else:
                    err_filename = orig_executable
                if errno_num != 0:
                    err_msg = os.strerror(errno_num)
>               raise child_exception_type(errno_num, err_msg, err_filename)
E               FileNotFoundError: [Errno 2] No such file or directory: 'exit'

/usr/lib/python3.10/subprocess.py:1845: FileNotFoundError
=================================================================== short test summary info ===================================================================
FAILED exit-test.py::test_assert_good_exitcode - FileNotFoundError: [Errno 2] No such file or directory: 'exit'
====================================================================== 1 failed in 0.05s ======================================================================
```

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -115,7 +115,7 @@ repos:
         name: Run bandit against the code base
         args: [--silent, -lll, --skip, B701]
         files: ^(?!tests/).*\.py$
-        exclude: src/pytestshellutils/version.py
+        exclude: ^src/pytestshellutils/(version|downgraded/shell)\.py$
   - repo: https://github.com/PyCQA/bandit
     rev: "1.7.4"
     hooks:

--- a/changelog/32.bugfix.rst
+++ b/changelog/32.bugfix.rst
@@ -1,0 +1,1 @@
+``Subprocess.run()`` now accepts ``shell`` keyword argument like ``subprocess.Popen``.

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -47,3 +47,8 @@ title_format = "shell-utilities {version} ({project_date})"
   directory = "trivial"
   name = "Trivial/Internal Changes"
   showcontent = true
+
+[tool.bandit]
+exclude = [
+  "src/pytestshellutils/downgraded/shell.py"
+]

--- a/src/pytestshellutils/downgraded/shell.py
+++ b/src/pytestshellutils/downgraded/shell.py
@@ -114,6 +114,7 @@ class SubprocessImpl:
     def init_terminal(
         self,
         cmdline: List[str],
+        shell: bool = False,
         env: Optional[EnvironDict] = None,
         cwd: Optional[Union[str, pathlib.Path]] = None,
     ) -> 'subprocess.Popen[Any]':
@@ -129,6 +130,8 @@ class SubprocessImpl:
                 List of strings to pass as ``args`` to :py:class:`~subprocess.Popen`
 
         Keyword Arguments:
+            shell:
+                Pass the value of ``shell`` to :py:class:`~subprocess.Popen`
             env:
                 A dictionary of ``key``, ``value`` pairs to add to the
                 :py:attr:`pytestshellutils.shell.Factory.environ`.
@@ -154,7 +157,7 @@ class SubprocessImpl:
             cmdline,
             stdout=self._terminal_stdout,
             stderr=self._terminal_stderr,
-            shell=False,
+            shell=shell,
             cwd=str(cwd or self.factory.cwd),
             universal_newlines=True,
             close_fds=close_fds,
@@ -281,6 +284,7 @@ class SubprocessImpl:
     def run(
         self,
         *args: str,
+        shell: bool = False,
         env: Optional[EnvironDict] = None,
         cwd: Optional[Union[str, pathlib.Path]] = None,
         **kwargs: Any
@@ -293,6 +297,9 @@ class SubprocessImpl:
                 The command to run.
 
         Keyword Arguments:
+            shell:
+                Pass the value of `shell` to
+                :py:meth:`pytestshellutils.shell.Factory.init_terminal`
             env:
                 A dictionary of ``key``, ``value`` pairs to add to the
                 :py:attr:`pytestshellutils.shell.Factory.environ`.
@@ -309,7 +316,7 @@ class SubprocessImpl:
             cmdline,
             cwd or self.factory.cwd,
         )
-        return self.init_terminal(cmdline, env=env, cwd=cwd)
+        return self.init_terminal(cmdline, shell=shell, env=env, cwd=cwd)
 
 
 @attr.s(slots=True, kw_only=True)

--- a/src/pytestshellutils/shell.py
+++ b/src/pytestshellutils/shell.py
@@ -121,6 +121,7 @@ class SubprocessImpl:
     def init_terminal(
         self,
         cmdline: List[str],
+        shell: bool = False,
         env: Optional[EnvironDict] = None,
         cwd: Optional[Union[str, pathlib.Path]] = None,
     ) -> "subprocess.Popen[Any]":
@@ -136,6 +137,8 @@ class SubprocessImpl:
                 List of strings to pass as ``args`` to :py:class:`~subprocess.Popen`
 
         Keyword Arguments:
+            shell:
+                Pass the value of ``shell`` to :py:class:`~subprocess.Popen`
             env:
                 A dictionary of ``key``, ``value`` pairs to add to the
                 :py:attr:`pytestshellutils.shell.Factory.environ`.
@@ -164,7 +167,7 @@ class SubprocessImpl:
             cmdline,
             stdout=self._terminal_stdout,
             stderr=self._terminal_stderr,
-            shell=False,
+            shell=shell,  # nosec B602
             cwd=str(cwd or self.factory.cwd),
             universal_newlines=True,
             close_fds=close_fds,
@@ -320,6 +323,7 @@ class SubprocessImpl:
     def run(
         self,
         *args: str,
+        shell: bool = False,
         env: Optional[EnvironDict] = None,
         cwd: Optional[Union[str, pathlib.Path]] = None,
         **kwargs: Any,
@@ -332,6 +336,9 @@ class SubprocessImpl:
                 The command to run.
 
         Keyword Arguments:
+            shell:
+                Pass the value of `shell` to
+                :py:meth:`pytestshellutils.shell.Factory.init_terminal`
             env:
                 A dictionary of ``key``, ``value`` pairs to add to the
                 :py:attr:`pytestshellutils.shell.Factory.environ`.
@@ -343,7 +350,7 @@ class SubprocessImpl:
         """
         cmdline = self.cmdline(*args, **kwargs)
         log.info("%s is running %r in CWD: %s ...", self.factory, cmdline, cwd or self.factory.cwd)
-        return self.init_terminal(cmdline, env=env, cwd=cwd)
+        return self.init_terminal(cmdline, shell=shell, env=env, cwd=cwd)
 
 
 @attr.s(slots=True, kw_only=True)

--- a/tests/functional/shell/test_fixture.py
+++ b/tests/functional/shell/test_fixture.py
@@ -5,6 +5,8 @@ import pathlib
 import sys
 import tempfile
 
+import pytest
+
 from pytestshellutils.shell import Subprocess
 from tests.conftest import Tempfiles
 
@@ -37,3 +39,12 @@ def test_run_cwd(tempfiles: Tempfiles, shell: Subprocess) -> None:
     result = shell.run(sys.executable, script, cwd=system_tempdir)
     assert result.returncode == 0
     assert result.stdout.strip() == system_tempdir
+
+
+def test_run_shell(shell: Subprocess) -> None:
+    with pytest.raises(FileNotFoundError):
+        shell.run("exit", "0")
+    with pytest.raises(FileNotFoundError):
+        shell.run("exit", "0", shell=False)
+    result = shell.run("exit", "0", shell=True)
+    assert result.returncode == 0

--- a/tests/functional/shell/test_subprocess.py
+++ b/tests/functional/shell/test_subprocess.py
@@ -228,3 +228,13 @@ def test_run_cwd(tmp_path: str, tempfiles: Tempfiles, shell: Subprocess) -> None
     result = shell.run(sys.executable, script, cwd=system_tempdir)
     assert result.returncode == 0
     assert result.stdout.strip() == system_tempdir
+
+
+def test_run_shell() -> None:
+    shell = Subprocess()
+    with pytest.raises(FileNotFoundError):
+        shell.run("exit", "0")
+    with pytest.raises(FileNotFoundError):
+        shell.run("exit", "0", shell=False)
+    result = shell.run("exit", "0", shell=True)
+    assert result.returncode == 0

